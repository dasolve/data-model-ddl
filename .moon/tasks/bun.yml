fileGroups:
  sources:
    - package.json
    - src/**/*
    - "!**/*.snap.png"
    - "**/*.json"
    - "**/*.ts"
    - "**/*.tsx"
    - "**/*.css"
    - "**/*.sql"
  eslint:
    - /packages/library-eslint/eslint.config.mjs
    - /.oxlintrc.json
    - /tsconfig.options.json
  not-storybook-files:
    - "!**/*.stories.{js,jsx,mjs,ts,tsx}"
    - "!.storybook/**"
  test-files:
    - "**/*.test.{js,jsx,mjs,ts,tsx}"
    - "**/mock.{ts,tsx}"
  not-test-files:
    - "!**/*.test.{js,jsx,mjs,ts,tsx}"
    - "!**/mock.{ts,tsx}"
  design-system:
    - "/packages/library-design-system/**/*.{ts,tsx}"
    - "!/packages/library-design-system/.storybook/**/*"
  ui-components:
    - "/packages/library-ui-components/**/*.{ts,tsx}"
    - "!/packages/library-ui-components/.storybook/**/*"
tasks:
  typecheck:
    toolchain: bun
    command: bun --bun tsc --build
    inputs:
      - "@group(sources)"
      - /tsconfig.options.json
      - /tsconfig.json
      - tsconfig.json
    deps:
      - "^:typecheck"
    type: test
    options:
      runInCI: true
      persistent: false
      mutex: lint
      outputStyle: buffer-only-failure
  oxlint:
    toolchain: bun
    command: bun --bun oxlint
    inputs:
      - "@group(sources)"
      - "@group(eslint)"
    type: test
    options:
      runInCI: true
      persistent: false
      internal: true
      cache: true
      outputStyle: buffer-only-failure
  # lint:
  #   toolchain: bun
  #   command: bun --bun eslint . --report-unused-disable-directives --max-warnings=0
  #   inputs:
  #     - "@group(sources)"
  #     - "@group(eslint)"
  #   deps:
  #     - oxlint
  #     - "^:lint"
  #   type: test
  #   options:
  #     cache: true
  #     timeout: 660
  #     runInCI: true
  #     persistent: false
  #     mutex: lint
  #     outputStyle: buffer-only-failure
  test:
    toolchain: bun
    command: bun test --coverage --coverage-reporter=lcov --coverage-dir=../../coverage/$project --reporter=junit --reporter-outfile=junit.xml
    inputs:
      - "@group(sources)"
    outputs:
      - junit.xml
      - /coverage/$project/lcov.info
    type: test
    deps:
      - "^:test"
    options:
      runInCI: true
      persistent: false
      # This is to avoid scenario where errors are not caught by moon
      timeout: 210
      retryCount: 2
      outputStyle: buffer-only-failure
  test-update-snapshots:
    toolchain: bun
    command: bun test --update-snapshots
    type: build
    local: true
    inputs:
      - "@group(test-files)"
    outputs:
      - "**/__snapshots__/*.snap"
    options:
      persistent: false
      runInCI: false
  format:
    toolchain: bun
    command: prettier --config=../../.prettierrc.json --ignore-path=../../.prettierignore --write .
    inputs:
      - "@group(sources)"
      - /.prettierrc.json
      - /.prettierignore
    type: run
    local: true
    options:
      persistent: false
  format-check:
    toolchain: bun
    command: prettier --config=../../.prettierrc.json --ignore-path=../../.prettierignore --check .
    inputs:
      - "@group(sources)"
      - /.prettierrc.json
      - /.prettierignore
    type: test
    options:
      runInCI: true
      persistent: false
      outputStyle: buffer-only-failure
