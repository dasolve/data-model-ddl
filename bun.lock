{
  "lockfileVersion": 1,
  "workspaces": {
    "": {
      "name": "data-model-dsl",
      "devDependencies": {
        "@types/bun": "catalog:",
        "oxlint": "1.7.0",
        "prettier": "catalog:",
        "typescript": "catalog:",
      },
    },
    "packages/drizzle": {
      "name": "@dasolve/dmddl-drizzle",
      "bin": {
        "dmddl-drizzle": "src/cli.ts",
      },
      "dependencies": {
        "@dasolve/dmddl-schema": "workspace:*",
        "yaml": "^2.8.0",
        "zod": "catalog:",
      },
      "peerDependencies": {
        "typescript": "catalog:",
      },
    },
    "packages/schema": {
      "name": "@dasolve/dmddl-schema",
      "dependencies": {
        "zod": "catalog:",
      },
      "peerDependencies": {
        "typescript": "catalog:",
      },
    },
  },
  "catalog": {
    "@types/bun": "latest",
    "prettier": "3.5.3",
    "typescript": "^5",
    "zod": "3.25.64",
  },
  "packages": {
    "@dasolve/dmddl-drizzle": ["@dasolve/dmddl-drizzle@workspace:packages/drizzle"],

    "@dasolve/dmddl-schema": ["@dasolve/dmddl-schema@workspace:packages/schema"],

    "@oxlint/darwin-arm64": ["@oxlint/darwin-arm64@1.7.0", "", { "os": "darwin", "cpu": "arm64" }, "sha512-51vhCSQO4NSkedwEwOyqThiYqV0DAUkwNdqMQK0d29j5zmtNJJJRRBLeQuLGdstNmn3F7WMQ75Ci0/3Nq4ff8A=="],

    "@oxlint/darwin-x64": ["@oxlint/darwin-x64@1.7.0", "", { "os": "darwin", "cpu": "x64" }, "sha512-c0GN52yehYZ4TYuh4lBH9wYbBOI/RDOxZhJdBsttG0GwfvKYg/tiPNrNEsPzu0/rd1j6x3yT0zt6vezDMeC1sQ=="],

    "@oxlint/linux-arm64-gnu": ["@oxlint/linux-arm64-gnu@1.7.0", "", { "os": "linux", "cpu": "arm64" }, "sha512-pam/lbzbzVMDzc3f1hoRPtnUMEIqkn0dynlB5nUll/MVBSIvIPLS9kJLrRA48lrlqbkS9LGiF37JvpwXA58A9A=="],

    "@oxlint/linux-arm64-musl": ["@oxlint/linux-arm64-musl@1.7.0", "", { "os": "linux", "cpu": "arm64" }, "sha512-LTyPy9FYS3SZ2XxJx+ITvlAq/ek5PtZK9Z2m3W72TA8hchGhJy5eQ+aotYjd/YVXOpGRpB12RdOpOTsZRu50bA=="],

    "@oxlint/linux-x64-gnu": ["@oxlint/linux-x64-gnu@1.7.0", "", { "os": "linux", "cpu": "x64" }, "sha512-YtZ4DiAgjaEiqUiwnvtJ/znZMAAVPKR7pnsi6lqbA3BfXJ/IwMaNpdoGlCGVdDGeN4BuGCwnFtBVqKVvVg3DDg=="],

    "@oxlint/linux-x64-musl": ["@oxlint/linux-x64-musl@1.7.0", "", { "os": "linux", "cpu": "x64" }, "sha512-5aIpemNUBvwMMk4MCx1V3M6R9eMB1/SS6/24Orax9FqaI1lDX08tySdv696sr4Lms9ocA+rotxIPW9NP9439vA=="],

    "@oxlint/win32-arm64": ["@oxlint/win32-arm64@1.7.0", "", { "os": "win32", "cpu": "arm64" }, "sha512-fpFpkHwbAu0NcR5bc1WapCPcM9qSYi5lCRVOp1WwDoFLKI2b9/UWB8OEg8UHWV5dnBu7HZAWH/SEslYGkZNsbQ=="],

    "@oxlint/win32-x64": ["@oxlint/win32-x64@1.7.0", "", { "os": "win32", "cpu": "x64" }, "sha512-0EPWBWOiD3wZHgeWDlTUaiFzhzIonXykxYUC+NRerPQFkO/G+bd9uLMJddHDKqfP/7g8s3E5V6KvBvvFpb7U6g=="],

    "@types/bun": ["@types/bun@1.2.19", "", { "dependencies": { "bun-types": "1.2.19" } }, "sha512-d9ZCmrH3CJ2uYKXQIUuZ/pUnTqIvLDS0SK7pFmbx8ma+ziH/FRMoAq5bYpRG7y+w1gl+HgyNZbtqgMq4W4e2Lg=="],

    "@types/node": ["@types/node@24.0.3", "", { "dependencies": { "undici-types": "~7.8.0" } }, "sha512-R4I/kzCYAdRLzfiCabn9hxWfbuHS573x+r0dJMkkzThEa7pbrcDWK+9zu3e7aBOouf+rQAciqPFMnxwr0aWgKg=="],

    "@types/react": ["@types/react@19.1.8", "", { "dependencies": { "csstype": "^3.0.2" } }, "sha512-AwAfQ2Wa5bCx9WP8nZL2uMZWod7J7/JSplxbTmBQ5ms6QpqNYm672H0Vu9ZVKVngQ+ii4R/byguVEUZQyeg44g=="],

    "bun-types": ["bun-types@1.2.19", "", { "dependencies": { "@types/node": "*" }, "peerDependencies": { "@types/react": "^19" } }, "sha512-uAOTaZSPuYsWIXRpj7o56Let0g/wjihKCkeRqUBhlLVM/Bt+Fj9xTo+LhC1OV1XDaGkz4hNC80et5xgy+9KTHQ=="],

    "csstype": ["csstype@3.1.3", "", {}, "sha512-M1uQkMl8rQK/szD0LNhtqxIPLpimGm8sOBwU7lLnCpSbTyY3yeU1Vc7l4KT5zT4s/yOxHH5O7tIuuLOCnLADRw=="],

    "oxlint": ["oxlint@1.7.0", "", { "optionalDependencies": { "@oxlint/darwin-arm64": "1.7.0", "@oxlint/darwin-x64": "1.7.0", "@oxlint/linux-arm64-gnu": "1.7.0", "@oxlint/linux-arm64-musl": "1.7.0", "@oxlint/linux-x64-gnu": "1.7.0", "@oxlint/linux-x64-musl": "1.7.0", "@oxlint/win32-arm64": "1.7.0", "@oxlint/win32-x64": "1.7.0" }, "bin": { "oxlint": "bin/oxlint", "oxc_language_server": "bin/oxc_language_server" } }, "sha512-krJN1fIRhs3xK1FyVyPtYIV9tkT4WDoIwI7eiMEKBuCjxqjQt5ZemQm1htPvHqNDOaWFRFt4btcwFdU8bbwgvA=="],

    "prettier": ["prettier@3.5.3", "", { "bin": { "prettier": "bin/prettier.cjs" } }, "sha512-QQtaxnoDJeAkDvDKWCLiwIXkTgRhwYDEQCghU9Z6q03iyek/rxRh/2lC3HB7P8sWT2xC/y5JDctPLBIGzHKbhw=="],

    "typescript": ["typescript@5.8.3", "", { "bin": { "tsc": "bin/tsc", "tsserver": "bin/tsserver" } }, "sha512-p1diW6TqL9L07nNxvRMM7hMMw4c5XOo/1ibL4aAIGmSAt9slTE1Xgw5KWuof2uTOvCg9BY7ZRi+GaF+7sfgPeQ=="],

    "undici-types": ["undici-types@7.8.0", "", {}, "sha512-9UJ2xGDvQ43tYyVMpuHlsgApydB8ZKfVYTsLDhXkFL/6gfkp+U8xTGdh8pMJv1SpZna0zxG1DwsKZsreLbXBxw=="],

    "yaml": ["yaml@2.8.0", "", { "bin": { "yaml": "bin.mjs" } }, "sha512-4lLa/EcQCB0cJkyts+FpIRx5G/llPxfP6VQU5KByHEhLxY3IJCH0f0Hy1MHI8sClTvsIb8qwRJ6R/ZdlDJ/leQ=="],

    "zod": ["zod@3.25.64", "", {}, "sha512-hbP9FpSZf7pkS7hRVUrOjhwKJNyampPgtXKc3AN6DsWtoHsg2Sb4SQaS4Tcay380zSwd2VPo9G9180emBACp5g=="],
  }
}
