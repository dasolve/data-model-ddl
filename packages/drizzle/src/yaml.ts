import { parseDataModel } from "./parser";
import { generateDrizzleSchemaFile } from "./generator";
import { readFileSync } from "node:fs";
import { join } from "node:path";
import { parse } from "yaml";

export const generateSchemaYamlFile = (filePath: string) => {
  const fullPath = join(import.meta.dir, filePath);
  const fileContent = readFileSync(fullPath, "utf-8");
  const parsedContent = parse(fileContent);
  const parsedDataModel = parseDataModel(parsedContent);
  return generateDrizzleSchemaFile(parsedDataModel);
};
