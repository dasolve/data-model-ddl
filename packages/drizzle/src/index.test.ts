import { describe, expect, test } from "bun:test";
import { readFileSync } from "fs";
import { join } from "path";
import { parse as parseYAML } from "yaml";
import { generateSchemaFileContent } from "./index";

// Utility function to load a YAML fixture from the fixtures directory
function loadFixture(fixtureName: string) {
  const fixturePath = join(import.meta.dir, "../fixtures", fixtureName);
  const fixtureContent = readFileSync(fixturePath, "utf-8");
  return parseYAML(fixtureContent);
}

describe("Drizzle Schema Generator", () => {
  test("should generate schema for basic table", () => {
    const fixture = loadFixture("basic-table.yml");
    const result = generateSchemaFileContent(fixture);
    expect(result).toMatchSnapshot();
  });

  test("should generate schema with relationships", () => {
    const fixture = loadFixture("relationships.yml");
    const result = generateSchemaFileContent(fixture);
    expect(result).toMatchSnapshot();
  });

  test("should generate schema with various column types", () => {
    const fixture = loadFixture("column-types.yml");
    const result = generateSchemaFileContent(fixture);
    expect(result).toMatchSnapshot();
  });

  test("should generate schema with constraints", () => {
    const fixture = loadFixture("constraints.yml");
    const result = generateSchemaFileContent(fixture);
    expect(result).toMatchSnapshot();
  });

  test("should generate schema with constraints (generated always as)", () => {
    const fixture = loadFixture("constraints-pk.yml");
    const result = generateSchemaFileContent(fixture);
    expect(result).toMatchSnapshot();
  });

  test("should handle errors for invalid inputs", () => {
    const fixture = loadFixture("invalid.yml");
    expect(() => generateSchemaFileContent(fixture)).toThrow();
  });

  test("should handle errors for invalid dialects", () => {
    const fixture = loadFixture("invalid-dialect.yml");
    expect(() => generateSchemaFileContent(fixture)).toThrow();
  });

  test("should handle errors for invalid uuid defaults", () => {
    const fixture = loadFixture("invalid-uuid-default.yml");
    expect(() => generateSchemaFileContent(fixture)).toThrow();
  });
});
