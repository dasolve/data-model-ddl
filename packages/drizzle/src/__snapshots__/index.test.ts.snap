// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`Drizzle Schema Generator should generate schema for basic table 1`] = `
"import { pgSchema, uuid, text, timestamp } from "drizzle-orm/pg-core";

/** A simple data model with basic tables */
const basicTableExample = pgSchema("basic_table_example");

/** User information table */
export const users = basicTableExample.table("users", {
  id: uuid("id").primaryKey().defaultRandom(),
  username: text("username").notNull().unique(),
  email: text("email").notNull().unique(),
  created_at: timestamp("created_at", { withTimeZone: true }).defaultNow()
});
"
`;

exports[`Drizzle Schema Generator should generate schema with relationships 1`] = `
"import { pgSchema, uuid, text, timestamp } from "drizzle-orm/pg-core";

/** A data model demonstrating table relationships */
const relationshipsExample = pgSchema("relationships_example");

/** Department information */
export const departments = relationshipsExample.table("departments", {
  id: uuid("id").primaryKey().defaultRandom(),
  name: text("name").notNull().unique()
});

/** Employee information */
export const employees = relationshipsExample.table("employees", {
  id: uuid("id").primaryKey().defaultRandom(),
  name: text("name").notNull(),
  department_id: uuid("department_id").notNull().references(() => departments.id),
  manager_id: uuid("manager_id").references(() => employees.id)
});

/** Project information */
export const projects = relationshipsExample.table("projects", {
  id: uuid("id").primaryKey().defaultRandom(),
  name: text("name").notNull()
});

/** Junction table for many-to-many relationship */
export const employeeProjects = relationshipsExample.table("employee_projects", {
  employee_id: uuid("employee_id").notNull().references(() => employees.id),
  project_id: uuid("project_id").notNull().references(() => projects.id)
});
"
`;

exports[`Drizzle Schema Generator should generate schema with various column types 1`] = `
"import { pgSchema, uuid, text, timestamp, integer, date } from "drizzle-orm/pg-core";

/** A data model showcasing different column types */
const columnTypesExample = pgSchema("column_types_example");

/** Example table with various column types */
export const productCatalog = columnTypesExample.table("product_catalog", {
  id: uuid("id").primaryKey().defaultRandom(),
  name: text("name").notNull(),
  description: text("description"),
  price: integer("price").notNull(),
  stock_quantity: integer("stock_quantity").default(0).notNull(),
  is_available: text("is_available").default("Y").notNull(),
  created_at: timestamp("created_at", { withTimeZone: true }).defaultNow(),
  updated_at: timestamp("updated_at", { withTimeZone: true }),
  release_date: date("release_date")
});
"
`;

exports[`Drizzle Schema Generator should generate schema with constraints 1`] = `
"import { pgSchema, uuid, text, timestamp, integer, date } from "drizzle-orm/pg-core";

/** A data model showcasing different constraints */
const constraintsExample = pgSchema("constraints_example");

/** Product information with constraints */
export const products = constraintsExample.table("products", {
  id: uuid("id").primaryKey().defaultRandom(),
  sku: text("sku").notNull().unique(),
  name: text("name").notNull(),
  price: integer("price").notNull(),
  stock: integer("stock").default(0).notNull(),
  category_id: uuid("category_id").notNull().references(() => categories.id)
});

/** Product categories */
export const categories = constraintsExample.table("categories", {
  id: uuid("id").primaryKey().defaultRandom(),
  name: text("name").notNull().unique()
});
"
`;

exports[`Drizzle Schema Generator should generate schema with constraints (generated always as) 1`] = `
"import { pgSchema, uuid, text, timestamp, integer, date } from "drizzle-orm/pg-core";

/** A data model showcasing different constraints */
const constraintsExample = pgSchema("constraints_example");

/** Product information with constraints */
export const products = constraintsExample.table("products", {
  id: uuid("id").primaryKey().defaultRandom(),
  sku: text("sku").notNull().unique(),
  name: text("name").notNull(),
  price: integer("price").notNull(),
  stock: integer("stock").default(0).notNull(),
  category_id: uuid("category_id").notNull().references(() => categories.id)
});

/** Product categories */
export const categories = constraintsExample.table("categories", {
  id: integer("id").primaryKey().generatedAlwaysAsIdentity(),
  name: text("name").notNull().unique()
});
"
`;
